### PR TITLE
Automatic resize

### PR DESCRIPTION
**Pull request by jbcolme**
I wrote a button that when active the plots resize automatically if the amplitude of a signal increases. Plus fixed a bug on the previous pull requested: When the signals where constant no resize happened. The history may be a little messy, I don't know how you handle this. 

There will be two buttons on the top bar: a straight arrow which represents a single resize of the plots, and a refresh icon to represent the constant and automatic option. I did it that way because I didn't now which other icons to use without making my own or expending too much time looking for options, so I grabbed two from QStyle::SP_whatever.

Cheers.

### CODE REVIEW

**Comment by patmarion** [comment]
Thanks for posting!  I'll review this when I get a chance.

File: src/signal_scope/mainwindow.cpp [hunk 1]
```diff
@@ -313,6 +315,25 @@ void MainWindow::onResize()
   }
 }
 
+void MainWindow::onAutomaticResize()
+{
+  if (mInternal->ActionAutomaticResize->isChecked()){
+    foreach (PlotWidget* plot, mPlots)
+    {
+      plot->onResetYAxisScale();
+      plot->getScale();
```
Comment (patmarion) [inline src/signal_scope/mainwindow.cpp]: is this line to call getScale() necessary?

File: src/signal_scope/plotwidget.cpp [hunk 2]
```diff
@@ -240,6 +247,47 @@ double PlotWidget::timeWindow() const
...
+    if (nn){
+      tmpRange[0] = signalData->value(0).y();
+      tmpRange[1] = signalData->value(0).y();
+
+      for (size_t jj=1; jj<signalData->size(); jj++){
+        if (signalData->value(jj).y() > tmpRange[1])
+          tmpRange[1] = signalData->value(jj).y();
+        else if (signalData->value(jj).y() < tmpRange[0])
+          tmpRange[0] = signalData->value(jj).y();
+      }
```
Comment (patmarion) [inline src/signal_scope/plotwidget.cpp]: i'm not sure why this function needs to compute the range?  The signalData object has a function called computeBounds() which returns the y range.  can you use that?

File: src/signal_scope/plotwidget.cpp [hunk 3]
```diff
@@ -240,6 +247,47 @@ double PlotWidget::timeWindow() const
...
+      yRange[1] = tmpRange[1];
+      update = true;
+    }      
+  }
+  return update;
+}
+
+void PlotWidget::resetYAxisMaxScale()
+{
+  if (getScale())
```
Comment (patmarion) [inline src/signal_scope/plotwidget.cpp]: maybe the getScale() function should be removed, and the update logic is encapsulated into this function resetYAxisMaxScale().  Each call to resetYAxisMaxScale() will check for an increased range and reset if the range has increased since the last call to resetYAxisMaxScale().  The getScale() function doesn't need to be in the API of PlotWidget.

Also, perhaps the cached Y range should be reset inside the call to PlotWidget::clearHistory()?

File: src/signal_scope/signaldata.cpp [hunk 4]
```diff
@@ -185,6 +185,13 @@ QRectF SignalData::computeBounds()
     if (y > maxY) maxY = y;
   }
 
+  double marginMax = 1.1;
+  double marginMin = 1.1;
+  if (maxY < 0.0) marginMax = 0.9;
+  if (minY > 0.0) marginMin = 0.9;
+  maxY *= marginMax;
+  minY *= marginMin;
```
Comment (patmarion) [inline src/signal_scope/signaldata.cpp]: can you explain what the above code is doing?  It seems like it is inflating the Y bounds computed by this function?  Why is that required?  It seems like this function should return the true bounds, and the caller can decide whether or not to inflate the bounds if needed.

File: src/signal_scope/plotwidget.cpp (hunk 2 above)
Comment (jbcolme) [inline src/signal_scope/plotwidget.cpp]: Ok I'll use that. Didn't notice it.

File: src/signal_scope/plotwidget.cpp (hunk 3 above)
Comment (jbcolme) [inline src/signal_scope/plotwidget.cpp]: You are right. I'll do the modifications. That's clearer code.

File: src/signal_scope/signaldata.cpp (hunk 4 above)
Comment (jbcolme) [inline src/signal_scope/signaldata.cpp]: Yes, it is inflating the code. Somebody in the team wanted to inflate the bounds automatically. It should be removed.

**Comment by jbcolme** [comment]
I'll make the modifications tomorrow and let you know. Thanks for going over the code.

File: src/signal_scope/signaldata.cpp (hunk 4 above)
Comment (patmarion) [inline src/signal_scope/signaldata.cpp]: Ok, that makes sense.  I think adding a margin to the plot Y limits is a reasonable feature to add, but it could be handled by the plot widget instead of the individual SignalData objects.

**Comment by jbcolme** [comment]
I think I fixed all the issues. The branch history is a mess sorry about that. Would you guys be interested in modifying the CMakeLists files so that they can be compiled with/without catkin? some people in our team are using ROS a lot, others pronto, so they are asking me to be able to compile it with catkin optionally.

File: src/signal_scope/plotwidget.cpp [hunk 5]
```diff
@@ -104,6 +106,11 @@ PlotWidget::PlotWidget(PythonChannelSubscriberCollection* subscribers, QWidget *
   QTimer* labelUpdateTimer = new QTimer(this);
   this->connect(labelUpdateTimer, SIGNAL(timeout()), SLOT(updateSignalInfoLabel()));
   labelUpdateTimer->start(100);
+
+  rescalingTimer = new QTimer(this);
+  this->connect(rescalingTimer, SIGNAL(timeout()), SLOT(resetYAxisMaxScale()));
+  yRange[0] = FLT_MAX;
+  yRange[1] = FLT_MIN;
```
Comment (patmarion) [inline src/signal_scope/plotwidget.cpp]: Could you use:

std::numeric_limits<float>::max()
-std::numeric_limits<float>::max()

I think FLT_MIN is possibly a bug?  FLT_MIN is close to 0.0.

File: src/signal_scope/plotwidget.cpp [hunk 6]
```diff
@@ -7,6 +7,7 @@
 #include "signaldescription.h"
 #include "pythonchannelsubscribercollection.h"
 
+#include <float.h>
```
Comment (patmarion) [inline src/signal_scope/plotwidget.cpp]: see comment about std::numeric_limits.  You could remove this include and replace with:

`#include <limits>`

File: src/signal_scope/plotwidget.cpp [hunk 7]
```diff
@@ -289,6 +321,17 @@ void PlotWidget::onResetYAxisScale()
...
+      if (signalBounds.top() == signalBounds.bottom()){  // in case of a constant signal
+        if (signalBounds.bottom() >= 0.0)
+          signalBounds.setBottom(signalBounds.bottom()*1.1);
+        else
+          signalBounds.setBottom(signalBounds.bottom()*0.9);
+
+        if (signalBounds.top() >= 0.0)
+          signalBounds.setTop(signalBounds.top()*0.9);
+        else
+          signalBounds.setTop(signalBounds.top()*1.1);
```
Comment (patmarion) [inline src/signal_scope/plotwidget.cpp]: possibly a simpler way to express this-

double margin = std::abs(signalBounds.top()*0.1)
signalBounds.setTop(signalBounds.top() + margin);
signalBounds.setBottom(signalBounds.bottom() - margin);

File: src/signal_scope/plotwidget.cpp [hunk 8]
```diff
@@ -305,6 +348,15 @@ void PlotWidget::onResetYAxisScale()
...
     area = QRectF(-1, -1, 2, 2);
   }
+  else
+  {
+      double marginMax = 1.1;
+      double marginMin = 1.1;
+      if (area.y()+area.height() < 0.0) marginMax = 0.9;
+      if (area.y() > 0.0) marginMin = 0.9;
+      area.setY(area.y()*marginMin);
+      area.setHeight(area.height()*marginMax);
```
Comment (patmarion) [inline src/signal_scope/plotwidget.cpp]: what's this code doing?

File: src/signal_scope/mainwindow.ui [hunk 9]
```diff
@@ -27,10 +27,10 @@
     <property name="title">
      <string>File</string>
     </property>
-    <addaction name="ActionOpen"/>
+    <!-- <addaction name="ActionOpen"/> -->
```
Comment (patmarion) [inline src/signal_scope/mainwindow.ui]: is this removing the open command?  Maybe instead of removing it we can rename it to "Open JSON".  Perhaps the JSON stuff can be retired soon, but I think we're still using it a little bit right now.  Perhaps we could do this on a separate pull request.

**Comment by patmarion** [comment]
Looking good.  I made a few more inline comments.  Thanks for working on this, it's nice to have this feature!

File: src/signal_scope/plotwidget.cpp (hunk 8 above)
Comment (jbcolme) [inline src/signal_scope/plotwidget.cpp]: So the previous code (the one you found a simpler way to write) increases the bounds if a signal is constant. If after a while a signal becomes constant the area won't be valid. That means the area.united() operation won't happen, and that signal may be left out the plot. So I increased the bounds if `signalBounds.top() == signalBounds.bottom()`. But now that you asked, and I'm going over the code again, I'm doing practically the same thing after in this piece of code to make a margin on the plot. Just that (as you suggested a few comments ago) the margin should be added on the plot, and not on each signal.

I'm going to rewrite this in a simpler way.

File: src/signal_scope/mainwindow.ui (hunk 9 above)
Comment (jbcolme) [inline src/signal_scope/mainwindow.ui]: Yes I messed up the branches. I started to write you an email about this but then decided to wait and do this first. I thought I had removed it.

**Comment by jbcolme** [comment]
Take a look at plotwidget and let me know what you think. I think this is clearer. Maybe a little redundant, I'm not sure its the "perfect" way. But it checks if the signal is constant and adds some margin so that the area is valid and the union works, and after adds some margin to the whole plot.

File: src/signal_scope/plotwidget.cpp [hunk 10]
```diff
@@ -305,6 +343,12 @@ void PlotWidget::onResetYAxisScale()
   {
     area = QRectF(-1, -1, 2, 2);
   }
+  else
+  {
+    double margin = std::abs(area.top()*0.1);
```
Comment (patmarion) [inline src/signal_scope/plotwidget.cpp]: at this point in the code, i think the rectangle has a valid height, so perhaps the margin should be a percentage of the height, and not a percentage of the absolute coordinate of the top or bottom?

File: src/signal_scope/plotwidget.cpp [hunk 11]
```diff
@@ -240,6 +248,30 @@ double PlotWidget::timeWindow() const
   return this->d_plot->timeWindow();
 }
 
+void PlotWidget::resetYAxisMaxScale()
+{
+  bool update = false;
+  QRectF tmpRange;
+  for (int ii=0; ii<mSignalListWidget->count(); ii++){
+    QListWidgetItem* signalItem = mSignalListWidget->item(ii);
+    SignalHandler* signalHandler = signalItem->data(Qt::UserRole).value<SignalHandler*>();
```
Comment (patmarion) [inline src/signal_scope/plotwidget.cpp:257]: perhaps we should add:

if (!this->signalIsVisible(signalHandler)) continue;

File: src/signal_scope/plotwidget.cpp [hunk 12]
```diff
@@ -104,6 +107,11 @@ PlotWidget::PlotWidget(PythonChannelSubscriberCollection* subscribers, QWidget *
   QTimer* labelUpdateTimer = new QTimer(this);
   this->connect(labelUpdateTimer, SIGNAL(timeout()), SLOT(updateSignalInfoLabel()));
   labelUpdateTimer->start(100);
+
+  rescalingTimer = new QTimer(this);
+  this->connect(rescalingTimer, SIGNAL(timeout()), SLOT(resetYAxisMaxScale()));
+  yRange[0] = std::numeric_limits<double>::max();
+  yRange[1] = -std::numeric_limits<double>::max();
```
Comment (patmarion) [inline src/signal_scope/plotwidget.cpp]: small nitpick, can you change these lines to use  "this->yRange"

Also, I think maybe the yRange should be reset to numeric_limits in the function PlotWidget::clearHistory().

**Comment by patmarion** [comment]
I made a few more in-line comments.  Sorry for the long review, but it's getting quite close now, thanks!

**Comment by jbcolme** [comment]
Done. How about now?

**Comment by patmarion** [comment]
Thanks for developing this, it's a really nice addition!
